### PR TITLE
Add integration hydreon_rgxx for rain sensors by Hydreon

### PR DESCRIPTION
**Pull request by functionpointer**
# What does this implement/fix? 

Adds integration for rain sensors [RG-9](https://rainsensors.com/products/rg-9/) and [RG-15](https://rainsensors.com/products/rg-15/).

I have tested it with an RG-9 and an esp32-mini. I don't have an RG-15, but the protocol is the same so i decided to implement support for it as well.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1616

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
uart:
  rx_pin: GPIO16
  tx_pin: GPIO17
  baud_rate: 9600

# Example RG-9 entry
sensor:
  - platform: hydreon_rgxx
    model: "RG_9"
    update_interval: 1s
    moisture:
      name: "rain"
      expire_after: 30s	
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by functionpointer** [comment]
Any idea when this will be reviewed? The associated [docs pull request](https://github.com/esphome/esphome-docs/pull/1616) got flagged as stale.

Maybe @jesserockz could take a look?

**Comment by functionpointer** [comment]
Once again the [associated docs pull request](https://github.com/esphome/esphome-docs/pull/1616) got flagged as stale.
I don't mean to sound impatient, but this PR is now four months old and hasn't received any feedback. Maybe @OttoWinter could take a look?

File: esphome/components/hydreon_rgxx/sensor.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            ),
+            cv.Optional(CONF_MOISTURE): sensor.sensor_schema(
+                unit_of_measurement=UNIT_INTENSITY,
+                accuracy_decimals=0,
+                device_class=DEVICE_CLASS_HUMIDITY,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+        }
+    )
+    .extend(cv.polling_component_schema("1s"))
```
Comment (jesserockz) [inline esphome/components/hydreon_rgxx/sensor.py]: Please default this to 60s.
```suggestion
    .extend(cv.polling_component_schema("60s"))
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/hydreon_rgxx/sensor.py (hunk 1 above)
Comment (functionpointer) [reply esphome/components/hydreon_rgxx/sensor.py]: Oh right!

File: esphome/components/hydreon_rgxx/binary_sensor.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        cv.GenerateID(): cv.declare_id(HydreonRGxxBinarySensor),
+        cv.GenerateID(CONF_HYDREON_RGXX_ID): cv.use_id(HydreonRGxxComponent),
+        cv.Optional(CONF_TOO_COLD): binary_sensor.BINARY_SENSOR_SCHEMA.extend(
+            {
+                cv.Optional(
+                    CONF_DEVICE_CLASS,
+                    default=DEVICE_CLASS_COLD,
+                ): binary_sensor.device_class
+            }
+        ),
```
Comment (jesserockz) [inline esphome/components/hydreon_rgxx/binary_sensor.py]: I made this simpler a few weeks ago =)
```suggestion
        cv.Optional(CONF_TOO_COLD): binary_sensor.binary_sensor_schema(
            device_class=DEVICE_CLASS_COLD
        ),
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
I just realised that you need to wrap every piece of binary_sensor code in this otherwise someone who has no binary sensors, but does have the sensors will not be able to compile.

```c++
#ifdef USE_BINARY_SENSOR
...
#endif
```

Also add this include to `hydreon_rgxx.h`
```c++
#include "esphome/core/defines.h"
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**